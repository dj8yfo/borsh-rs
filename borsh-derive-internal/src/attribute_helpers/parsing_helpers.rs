// TODO: remove this unused attribute, when the unsplit is done
#![allow(unused)]
use std::iter::FromIterator;

use syn::{
    meta::ParseNestedMeta, punctuated::Punctuated, token::Paren, Expr, Ident, Lit, LitStr, Token,
    Type, WherePredicate,
};

use super::{Bounds, SchemaParams, Symbol, BOUND, DESERIALIZE, PARAMS, SCHEMA, SERIALIZE};
fn get_lit_str2(
    attr_name: Symbol,
    meta_item_name: Symbol,
    meta: &ParseNestedMeta,
) -> syn::Result<Option<LitStr>> {
    let expr: Expr = meta.value()?.parse()?;
    let mut value = &expr;
    while let Expr::Group(e) = value {
        value = &e.expr;
    }
    if let Expr::Lit(syn::ExprLit {
        lit: Lit::Str(lit), ..
    }) = value
    {
        Ok(Some(lit.clone()))
    } else {
        Err(syn::Error::new_spanned(
            expr,
            format!(
                "expected borsh {} attribute to be a string: `{} = \"...\"`",
                attr_name.0, meta_item_name.0
            ),
        ))
    }
}

fn parse_lit_into<T: syn::parse::Parse>(
    attr_name: Symbol,
    meta_item_name: Symbol,
    meta: &ParseNestedMeta,
) -> syn::Result<Vec<T>> {
    let string = match get_lit_str2(attr_name, meta_item_name, meta)? {
        Some(string) => string,
        None => return Ok(Vec::new()),
    };

    match string.parse_with(Punctuated::<T, Token![,]>::parse_terminated) {
        Ok(elements) => Ok(Vec::from_iter(elements)),
        Err(err) => Err(syn::Error::new_spanned(string, err)),
    }
}

/**
Struct describes an entry like `order_param => override_type`,  e.g. `K => <K as TraitName>::Associated`
*/
#[derive(Clone, syn_derive::Parse, syn_derive::ToTokens)]
pub struct SchemaParamsOverride {
    pub order_param: Ident,
    arrow_token: Token![=>],
    pub override_type: Type,
}

fn get_ser_and_de<T, F, R>(
    attr_name: Symbol,
    meta: &ParseNestedMeta,
    f: F,
) -> syn::Result<(Option<T>, Option<T>)>
where
    T: Clone,
    F: Fn(Symbol, Symbol, &ParseNestedMeta) -> syn::Result<R>,
    R: Into<Option<T>>,
{
    let mut ser_meta: Option<T> = None;
    let mut de_meta = None;

    let lookahead = meta.input.lookahead1();
    if lookahead.peek(Paren) {
        meta.parse_nested_meta(|meta| {
            if meta.path == SERIALIZE {
                if let Some(v) = f(attr_name, SERIALIZE, &meta)?.into() {
                    ser_meta = Some(v);
                }
            } else if meta.path == DESERIALIZE {
                if let Some(v) = f(attr_name, DESERIALIZE, &meta)?.into() {
                    de_meta = Some(v);
                }
            } else {
                return Err(meta.error(format_args!(
                    "malformed {0} attribute, expected `{0}(serialize = ..., deserialize = ...)`",
                    attr_name.0,
                )));
            }
            Ok(())
        })?;
    } else {
        return Err(lookahead.error());
    }

    Ok((ser_meta, de_meta))
}

fn get_schema_nested_meta<T, F, R>(
    attr_name: Symbol,
    meta: &ParseNestedMeta,
    f: F,
) -> syn::Result<Option<T>>
where
    T: Clone,
    F: Fn(Symbol, Symbol, &ParseNestedMeta) -> syn::Result<R>,
    R: Into<Option<T>>,
{
    let mut params: Option<T> = None;

    let lookahead = meta.input.lookahead1();
    if lookahead.peek(Paren) {
        meta.parse_nested_meta(|meta| {
            if meta.path == PARAMS {
                if let Some(v) = f(attr_name, PARAMS, &meta)?.into() {
                    params = Some(v);
                }
            } else {
                return Err(meta.error(format_args!(
                    "malformed {0} attribute, expected `{0}(params = ...)`",
                    attr_name.0,
                )));
            }
            Ok(())
        })?;
    } else {
        return Err(lookahead.error());
    }

    Ok(params)
}

pub(crate) fn get_where_predicates(meta: &ParseNestedMeta) -> syn::Result<(Bounds, Bounds)> {
    get_ser_and_de(BOUND, meta, parse_lit_into::<WherePredicate>)
}

pub(crate) fn get_schema_attrs(meta: &ParseNestedMeta) -> syn::Result<SchemaParams> {
    get_schema_nested_meta(SCHEMA, meta, parse_lit_into::<SchemaParamsOverride>)
}
