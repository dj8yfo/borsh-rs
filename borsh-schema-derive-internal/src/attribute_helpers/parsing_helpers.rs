// TODO: remove this unused attribute, when the unsplit is done
#![allow(unused)]
use std::iter::FromIterator;

use syn::{meta::ParseNestedMeta, punctuated::Punctuated, token::Paren, Expr, Lit, LitStr, Token};

use super::{Bounds, Symbol, BOUND, DESERIALIZE, SERIALIZE};
fn get_lit_str2(
    attr_name: Symbol,
    meta_item_name: Symbol,
    meta: &ParseNestedMeta,
) -> syn::Result<Option<LitStr>> {
    let expr: Expr = meta.value()?.parse()?;
    let mut value = &expr;
    while let Expr::Group(e) = value {
        value = &e.expr;
    }
    if let Expr::Lit(syn::ExprLit {
        lit: Lit::Str(lit), ..
    }) = value
    {
        Ok(Some(lit.clone()))
    } else {
        Err(syn::Error::new_spanned(
            expr,
            format!(
                "expected borsh {} attribute to be a string: `{} = \"...\"`",
                attr_name.0, meta_item_name.0
            ),
        ))
    }
}

fn parse_lit_into_where(
    attr_name: Symbol,
    meta_item_name: Symbol,
    meta: &ParseNestedMeta,
) -> syn::Result<Vec<syn::WherePredicate>> {
    let string = match get_lit_str2(attr_name, meta_item_name, meta)? {
        Some(string) => string,
        None => return Ok(Vec::new()),
    };

    match string.parse_with(Punctuated::<syn::WherePredicate, Token![,]>::parse_terminated) {
        Ok(predicates) => Ok(Vec::from_iter(predicates)),
        Err(err) => Err(syn::Error::new_spanned(string, err)),
    }
}

fn get_ser_and_de<T, F, R>(
    attr_name: Symbol,
    meta: &ParseNestedMeta,
    f: F,
) -> syn::Result<(Option<T>, Option<T>)>
where
    T: Clone,
    F: Fn(Symbol, Symbol, &ParseNestedMeta) -> syn::Result<R>,
    R: Into<Option<T>>,
{
    let mut ser_meta: Option<T> = None;
    let mut de_meta = None;

    let lookahead = meta.input.lookahead1();
    if lookahead.peek(Paren) {
        meta.parse_nested_meta(|meta| {
            if meta.path == SERIALIZE {
                if let Some(v) = f(attr_name, SERIALIZE, &meta)?.into() {
                    ser_meta = Some(v);
                }
            } else if meta.path == DESERIALIZE {
                if let Some(v) = f(attr_name, DESERIALIZE, &meta)?.into() {
                    de_meta = Some(v);
                }
            } else {
                return Err(meta.error(format_args!(
                    "malformed {0} attribute, expected `{0}(serialize = ..., deserialize = ...)`",
                    attr_name.0,
                )));
            }
            Ok(())
        })?;
    } else {
        return Err(lookahead.error());
    }

    Ok((ser_meta, de_meta))
}
pub fn get_where_predicates(meta: &ParseNestedMeta) -> syn::Result<(Bounds, Bounds)> {
    get_ser_and_de(BOUND, meta, parse_lit_into_where)
}