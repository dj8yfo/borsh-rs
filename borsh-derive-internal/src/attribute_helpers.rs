use syn::{Attribute, Field, Path, WherePredicate};
pub mod parsing_helpers;
use parsing_helpers::get_where_predicates;

#[derive(Copy, Clone)]
pub struct Symbol(pub &'static str);

pub const BORSH: Symbol = Symbol("borsh");
pub const BOUND: Symbol = Symbol("bound");
pub const SERIALIZE: Symbol = Symbol("serialize");
pub const DESERIALIZE: Symbol = Symbol("deserialize");

impl PartialEq<Symbol> for Path {
    fn eq(&self, word: &Symbol) -> bool {
        self.is_ident(word.0)
    }
}

impl<'a> PartialEq<Symbol> for &'a Path {
    fn eq(&self, word: &Symbol) -> bool {
        self.is_ident(word.0)
    }
}

pub fn contains_skip(attrs: &[Attribute]) -> bool {
    attrs.iter().any(|attr| attr.path().is_ident("borsh_skip"))
}

pub fn contains_initialize_with(attrs: &[Attribute]) -> Option<Path> {
    for attr in attrs.iter() {
        if attr.path().is_ident("borsh_init") {
            let mut res = None;
            let _ = attr.parse_nested_meta(|meta| {
                res = Some(meta.path);
                Ok(())
            });
            return res;
        }
    }

    None
}

type Bounds = Option<Vec<WherePredicate>>;

pub fn parse_bounds(attrs: &[Attribute]) -> Result<(Bounds, Bounds), syn::Error> {
    let (mut ser, mut de): (Bounds, Bounds) = (None, None);
    for attr in attrs {
        if attr.path() != BORSH {
            continue;
        }

        attr.parse_nested_meta(|meta| {
            if meta.path == BOUND {
                // #[borsh(bound(serialize = "...", deserialize = "...", schema = "..."))]

                let (ser_parsed, de_parsed) = get_where_predicates(&meta)?;
                ser = ser_parsed;
                de = de_parsed;
            }
            Ok(())
        })?;
    }

    Ok((ser, de))
}

pub enum BoundType {
    Serialize,
    Deserialize,
}

pub fn get_bounds(field: &Field, ty: BoundType) -> Result<Bounds, syn::Error> {
    let (ser, de) = parse_bounds(&field.attrs)?;
    match ty {
        BoundType::Serialize => Ok(ser),
        BoundType::Deserialize => Ok(de),
    }
}

pub fn collect_override_bounds(
    field: &Field,
    ty: BoundType,
    output: &mut Vec<WherePredicate>,
) -> Result<bool, syn::Error> {
    let predicates = get_bounds(field, ty)?;
    match predicates {
        Some(predicates) => {
            output.extend(predicates);
            Ok(true)
        }
        None => Ok(false),
    }
}

#[cfg(test)]
mod tests {
    use quote::{quote, ToTokens};
    use std::fmt::Write;
    use syn::ItemStruct;

    use super::{parse_bounds, Bounds};
    fn debug_print_bounds(bounds: Bounds) -> String {
        let mut s = String::new();
        if let Some(bounds) = bounds {
            for bound in bounds {
                writeln!(&mut s, "{}", bound.to_token_stream()).unwrap();
            }
        } else {
            write!(&mut s, "None").unwrap();
        }
        s
    }

    #[test]
    fn test_bounds_parsing1() {
        let item_struct: ItemStruct = syn::parse2(quote! {
            struct A {
                #[borsh(bound(deserialize = "K: Hash + Ord,
                     V: Eq + Ord",
                    serialize = "K: Hash + Eq + Ord,
                     V: Ord"
                ))]
                x: u64,
                y: String,
            }
        })
        .unwrap();

        let first_field = &item_struct.fields.into_iter().collect::<Vec<_>>()[0];
        let (ser, de) = parse_bounds(&first_field.attrs).unwrap();
        insta::assert_snapshot!(debug_print_bounds(ser));
        insta::assert_snapshot!(debug_print_bounds(de));
    }

    #[test]
    fn test_bounds_parsing2() {
        let item_struct: ItemStruct = syn::parse2(quote! {
            struct A {
                #[borsh(bound(deserialize = "K: Hash + Eq + borsh::de::BorshDeserialize,
                     V: borsh::de::BorshDeserialize",
                    serialize = "K: Hash + Eq + borsh::ser::BorshSerialize,
                     V: borsh::ser::BorshSerialize"
                ))]
                x: u64,
                y: String,
            }
        })
        .unwrap();

        let first_field = &item_struct.fields.into_iter().collect::<Vec<_>>()[0];
        let (ser, de) = parse_bounds(&first_field.attrs).unwrap();
        insta::assert_snapshot!(debug_print_bounds(ser));
        insta::assert_snapshot!(debug_print_bounds(de));
    }

    #[test]
    fn test_bounds_parsing3() {
        let item_struct: ItemStruct = syn::parse2(quote! {
            struct A {
                #[borsh(bound(deserialize = "K: Hash + Eq + borsh::de::BorshDeserialize,
                     V: borsh::de::BorshDeserialize",
                    serialize = ""
                ))]
                x: u64,
                y: String,
            }
        })
        .unwrap();

        let first_field = &item_struct.fields.into_iter().collect::<Vec<_>>()[0];
        let (ser, de) = parse_bounds(&first_field.attrs).unwrap();
        insta::assert_snapshot!(debug_print_bounds(ser));
        insta::assert_snapshot!(debug_print_bounds(de));
    }

    #[test]
    fn test_bounds_parsing4() {
        let item_struct: ItemStruct = syn::parse2(quote! {
            struct A {
                #[borsh(bound(deserialize = "K: Hash"))]
                x: u64,
                y: String,
            }
        })
        .unwrap();

        let first_field = &item_struct.fields.into_iter().collect::<Vec<_>>()[0];
        let (ser, de) = parse_bounds(&first_field.attrs).unwrap();
        insta::assert_snapshot!(debug_print_bounds(ser));
        insta::assert_snapshot!(debug_print_bounds(de));
    }

    #[test]
    fn test_bounds_parsing_error() {
        let item_struct: ItemStruct = syn::parse2(quote! {
            struct A {
                #[borsh(bound(deser = "K: Hash"))]
                x: u64,
                y: String,
            }
        })
        .unwrap();

        let first_field = &item_struct.fields.into_iter().collect::<Vec<_>>()[0];
        let err = match parse_bounds(&first_field.attrs) {
            Ok(..) => unreachable!("expecting error here"),
            Err(err) => err,
        };
        insta::assert_debug_snapshot!(err);
    }

    #[test]
    fn test_bounds_parsing_error2() {
        let item_struct: ItemStruct = syn::parse2(quote! {
            struct A {
                #[borsh(bound(deserialize = "K Hash"))]
                x: u64,
                y: String,
            }
        })
        .unwrap();

        let first_field = &item_struct.fields.into_iter().collect::<Vec<_>>()[0];
        let err = match parse_bounds(&first_field.attrs) {
            Ok(..) => unreachable!("expecting error here"),
            Err(err) => err,
        };
        insta::assert_debug_snapshot!(err);
    }

    #[test]
    fn test_bounds_parsing_error3() {
        let item_struct: ItemStruct = syn::parse2(quote! {
            struct A {
                #[borsh(bound(deserialize = 42))]
                x: u64,
                y: String,
            }
        })
        .unwrap();

        let first_field = &item_struct.fields.into_iter().collect::<Vec<_>>()[0];
        let err = match parse_bounds(&first_field.attrs) {
            Ok(..) => unreachable!("expecting error here"),
            Err(err) => err,
        };
        insta::assert_debug_snapshot!(err);
    }
}
