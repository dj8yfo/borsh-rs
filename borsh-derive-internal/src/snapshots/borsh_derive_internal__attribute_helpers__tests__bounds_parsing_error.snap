---
source: borsh-derive-internal/src/attribute_helpers.rs
expression: err
---
Error(
    "malformed bound attribute, expected `bound(serialize = ..., deserialize = ..., schema = ...)`",
)
